File: src/Application Framework/Structures/DataGridCombo/XmlPersister.vb
Imports System.IO
Imports System.Xml.Serialization

Public Class XmlPersister

#Region " accessors "

    Private Shared _defaultfolder As String = String.Empty

    Public Shared Property DefaultFolder() As String
        Get
            Return _defaultfolder
        End Get
        Set(ByVal Value As String)
            If Value.EndsWith("\bin") Then
                Value = Value.Remove(Value.Length - 3, 3)
            End If
            If Not Value.EndsWith("\") Then Value += "\"
            _defaultfolder = Value
        End Set
    End Property

#End Region

#Region " shared serialise methods "

    Public Shared Function DefaultListExits(ByVal objectType As Type) As Boolean
        Dim myName As String
        myName = String.Format("{0}{1}.xml", XmlPersister.DefaultFolder, objectType.Name)
        Return XmlPersister.Exists(myName)
    End Function

    Public Shared Function Exists(ByVal persistName As String)

        Try
            Dim myIO As New System.IO.FileStream(persistName, FileMode.Open)
            myIO.Close()
            Return True
        Catch ex As Exception
            If ex.GetType.Name = "FileNotFoundException" Then
                Return False
            End If
        End Try

    End Function

    Public Shared Function GetDefaultList(ByVal objectType As Type) As IPersistable
        Dim myName As String
        myName = String.Format("{0}{1}.xml", XmlPersister.DefaultFolder, objectType.Name)
        Return XmlPersister.Read(myName, objectType)
    End Function

    Public Shared Function Read(ByVal persistName As String, _
                                ByVal objectType As Type) As IPersistable

        Dim myObject As IPersistable

        Try

            Dim mySerializer As XmlSerializer = New XmlSerializer([objectType])
            Dim myStream As FileStream = New FileStream(persistName, FileMode.Open)

            Try

                myObject = mySerializer.Deserialize(myStream)
                myStream.Close()

                myObject.HasBeenPersisted = True
                myObject.PersistName = persistName

            Catch ex As Exception

                System.Diagnostics.Debug.Assert(False, ex.Message)

            End Try

        Catch ex As Exception

            System.Diagnostics.Debug.Assert(False, ex.Message)

        End Try

        Return myObject

    End Function

    Public Shared Function Write(ByVal persistableObject As IPersistable)

        Dim myObject As Object = persistableObject

        If persistableObject.PersistName = persistableObject.CurrentLocation Then

            persistableObject.PersistName = persistableObject.CurrentLocation + _
                                            persistableObject.DefaultName

        End If


        Dim mySerializer As XmlSerializer = New XmlSerializer(myObject.GetType())
        Dim myStream As StreamWriter = New StreamWriter(persistableObject.PersistName)

        Try

            mySerializer.Serialize(myStream, myObject)
            persistableObject.HasBeenPersisted = True

        Catch ex As Exception

            System.Diagnostics.Debug.Assert(False, ex.Message)

        End Try

        myStream.Close()

    End Function

#End Region

End Class

File: src/Application Framework/Structures/DataGridCombo/XmlPersister.cs
using Microsoft.VisualBasic.CompilerServices;
using System;
using System.Diagnostics;
using System.IO;
using System.Xml.Serialization;

namespace FSM
{
    public class XmlPersister
    {
        /* TODO ERROR: Skipped RegionDirectiveTrivia */
        private static string _defaultfolder = string.Empty;

        public static string DefaultFolder
        {
            get
            {
                return _defaultfolder;
            }

            set
            {
                if (value.EndsWith(@"\bin"))
                {
                    value = value.Remove(value.Length - 3, 3);
                }

                if (!value.EndsWith(@"\"))
                    value += @"\";
                _defaultfolder = value;
            }
        }

        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
        /* TODO ERROR: Skipped RegionDirectiveTrivia */

        public static bool DefaultListExits(Type objectType)
        {
            string myName;
            myName = string.Format("{0}{1}.xml", DefaultFolder, objectType.Name);
            return Conversions.ToBoolean(Exists(myName));
        }

        public static object Exists(string persistName)
        {
            try
            {
                var myIO = new FileStream(persistName, FileMode.Open);
                myIO.Close();
                return true;
            }
            catch (Exception ex)
            {
                if ((ex.GetType().Name ?? "") == "FileNotFoundException")
                {
                    return false;
                }
            }

            return default;
        }

        public static Interfaces.IPersistable GetDefaultList(Type objectType)
        {
            string myName;
            myName = string.Format("{0}{1}.xml", DefaultFolder, objectType.Name);
            return Read(myName, objectType);
        }

        public static Interfaces.IPersistable Read(string persistName, Type objectType)
        {
            var myObject = default(Interfaces.IPersistable);
            try
            {
                var mySerializer = new XmlSerializer(objectType);
                var myStream = new FileStream(persistName, FileMode.Open);
                try
                {
                    myObject = (Interfaces.IPersistable)mySerializer.Deserialize(myStream);
                    myStream.Close();
                    myObject.HasBeenPersisted = true;
                    myObject.PersistName = persistName;
                }
                catch (Exception ex)
                {
                    Debug.Assert(false, ex.Message);
                }
            }
            catch (Exception ex)
            {
                Debug.Assert(false, ex.Message);
            }

            return myObject;
        }

        public static object Write(Interfaces.IPersistable persistableObject)
        {
            object myObject = persistableObject;
            if ((persistableObject.PersistName ?? "") == (persistableObject.CurrentLocation ?? ""))
            {
                persistableObject.PersistName = persistableObject.CurrentLocation + persistableObject.DefaultName;
            }

            var mySerializer = new XmlSerializer(myObject.GetType());
            var myStream = new StreamWriter(persistableObject.PersistName);
            try
            {
                mySerializer.Serialize(myStream, myObject);
                persistableObject.HasBeenPersisted = true;
            }
            catch (Exception ex)
            {
                Debug.Assert(false, ex.Message);
            }

            myStream.Close();
            return default;
        }

        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
    }
}